File: Tests/CSharp/src/Tests/SettingsSystemTests.cs
using System.IO;
using System.Runtime.InteropServices;
using Common = Csp.Common;
using Services = Csp.Services;
using Systems = Csp.Systems;

using CSharpTests;
using static CSharpTests.TestHelper;


namespace CSPEngine
{
    static class SettingsSystemTests
    {
#if RUN_ALL_UNIT_TESTS || RUN_SETTINGSSYSTEM_TESTS || RUN_SETTINGSSYSTEM_BLOCKSPACE_TEST
        [Test]
        public static void BlockSpaceTest()
        {
            GetFoundationSystems(out var userSystem, out var spaceSystem, out _, out _, out _, out _, out var settingsSystem, out _);

            string testSpaceName = GenerateUniqueString("OLY-UNITTEST-SPACE-REWIND");
            string testSpaceDescription = "OLY-UNITTEST-SPACEDESC-REWIND";

            // Log in
            var userId = UserSystemTests.LogIn(userSystem);

            // Create space
            var space = SpaceSystemTests.CreateSpace(spaceSystem, testSpaceName, testSpaceDescription, Systems.SpaceAttributes.Private, null, null, null);

            // Clear blocked spaces
            {
                using var result = settingsSystem.ClearBlockedSpaces(userId).Result;
                var resCode = result.GetResultCode();

                Assert.AreEqual(resCode, Services.EResultCode.Success);
            }

            // Block space
            {
                using var result = settingsSystem.AddBlockedSpace(userId, space.Id).Result;
                var resCode = result.GetResultCode();

                Assert.AreEqual(resCode, Services.EResultCode.Success);
            }

            // Get blocked spaces
            {
                using var result = settingsSystem.GetBlockedSpaces(userId).Result;
                var resCode = result.GetResultCode();

                Assert.AreEqual(resCode, Services.EResultCode.Success);

                using var blockedSpaces = result.GetValue();

                Assert.AreEqual(blockedSpaces.Size(), 1UL);
                Assert.AreEqual(blockedSpaces[0], space.Id);
            }

            // Clear blocked spaces
            {
                using var result = settingsSystem.ClearBlockedSpaces(userId).Result;
                var resCode = result.GetResultCode();

                Assert.AreEqual(resCode, Services.EResultCode.Success);
            }
        }
#endif

#if RUN_ALL_UNIT_TESTS || RUN_SETTINGSSYSTEM_TESTS || RUN_SETTINGSSYSTEM_UNBLOCKSPACE_TEST
        [Test]
        public static void UnblockSpaceTest()
        {
            GetFoundationSystems(out var userSystem, out var spaceSystem, out _, out _, out _, out _, out var settingsSystem, out _);

            string testSpaceName = GenerateUniqueString("OLY-UNITTEST-SPACE-REWIND");
            string testSpaceDescription = "OLY-UNITTEST-SPACEDESC-REWIND";

            // Log in
            var userId = UserSystemTests.LogIn(userSystem);

            // Create space
            var space = SpaceSystemTests.CreateSpace(spaceSystem, testSpaceName, testSpaceDescription, Systems.SpaceAttributes.Private, null, null, null);

            // Clear blocked spaces
            {
                using var result = settingsSystem.ClearBlockedSpaces(userId).Result;
                var resCode = result.GetResultCode();

                Assert.AreEqual(resCode, Services.EResultCode.Success);
            }

            // Block space
            {
                using var result = settingsSystem.AddBlockedSpace(userId, space.Id).Result;
                var resCode = result.GetResultCode();

                Assert.AreEqual(resCode, Services.EResultCode.Success);
            }

            // Get blocked spaces
            {
                using var result = settingsSystem.GetBlockedSpaces(userId).Result;
                var resCode = result.GetResultCode();

                Assert.AreEqual(resCode, Services.EResultCode.Success);

                using var blockedSpaces = result.GetValue();

                Assert.AreEqual(blockedSpaces.Size(), 1UL);
                Assert.AreEqual(blockedSpaces[0], space.Id);
            }

            // Remove Blocked Spaces
            {
                using var result = settingsSystem.RemoveBlockedSpace(userId, space.Id).Result;
                var resCode = result.GetResultCode();

                Assert.AreEqual(resCode, Services.EResultCode.Success);
            }

            // Get blocked spaces
            {
                using var result = settingsSystem.GetBlockedSpaces(userId).Result;
                var resCode = result.GetResultCode();

                Assert.AreEqual(resCode, Services.EResultCode.Success);

                using var blockedSpaces = result.GetValue();

                Assert.AreEqual(blockedSpaces.Size(), 0UL);
            }

            // Clear blocked spaces
            {
                using var result = settingsSystem.ClearBlockedSpaces(userId).Result;
                var resCode = result.GetResultCode();

                Assert.AreEqual(resCode, Services.EResultCode.Success);
            }
        }
#endif
#if RUN_ALL_UNIT_TESTS || RUN_SETTINGSSYSTEM_TESTS || RUN_SETTINGSSYSTEM_GETAVATAR_PORTRAIT_TEST
        [Test]
        public static void GetAvatarPortraitTest()
        {
            GetFoundationSystems(out var userSystem, out var spaceSystem, out _, out _, out _, out _, out var settingsSystem, out _);

            var testSpaceName = GenerateUniqueString("OLY-UNITTEST-SPACE-REWIND");
            var testSpaceDescription = "OLY-UNITTEST-SPACEDESC-REWIND";

            var userId = userSystem.TestLogIn(pushCleanupFunction: false);

            var localFileName = "OKO.png";
            var AvatarPortrait = new Systems.FileAssetDataSource
            {
                FilePath = Path.GetFullPath("assets/" + localFileName)
            };
            {
                using var result = settingsSystem.UpdateAvatarPortrait(userId, AvatarPortrait).Result;

                Assert.AreEqual(result.GetResultCode(), Services.EResultCode.Success);
            }
            {
                using var result = settingsSystem.GetAvatarPortrait(userId).Result;

                Assert.AreEqual(result.GetResultCode(), Services.EResultCode.Success);
                Assert.IsTrue(result.GetUri().Contains(localFileName));
            }

            UserSystemTests.LogOut(userSystem);
        }
#endif
#if RUN_ALL_UNIT_TESTS || RUN_SETTINGSSYSTEM_TESTS || RUN_SETTINGSSYSTEM_UPDATEAVATAR_PORTRAIT_WITH_BUFFER_TEST
        [Test]
        public static void UpdateAvatarPortraitWithBufferTest()
        {
            GetFoundationSystems(out var userSystem, out var spaceSystem, out var assetSystem, out _, out _, out _, out var settingsSystem, out _);
            var userId = userSystem.TestLogIn(email: UserSystemTests.AlternativeLoginEmail, password: UserSystemTests.AlternativeLoginPassword);

            {
                var uploadFilePath = Path.GetFullPath("assets/OKO.png");
                var uploadFileSize = new System.IO.FileInfo(uploadFilePath).Length;

                var uploadFileData = File.ReadAllBytes(uploadFilePath);

                System.IntPtr uploadFileDataPtr = Marshal.AllocHGlobal((int)uploadFileSize);
                Marshal.Copy(uploadFileData, 0, uploadFileDataPtr, (int)uploadFileSize);

                var source = new Systems.BufferAssetDataSource
                {
                    Buffer = uploadFileDataPtr,
                    BufferLength = (ulong)uploadFileSize
                };
                source.SetMimeType("image/png");

                using var result = settingsSystem.UpdateAvatarPortraitWithBuffer(userId, source).Result;

                Assert.AreEqual(result.GetResultCode(), Services.EResultCode.Success);

                using var resultGetThumbnail = settingsSystem.GetAvatarPortrait(userId).Result;

                Assert.AreEqual(resultGetThumbnail.GetResultCode(), Services.EResultCode.Success);

                var downloadedAsset = new Systems.Asset
                {
                    Uri = resultGetThumbnail.GetUri()
                };

                using var downloadResult = assetSystem.DownloadAssetData(downloadedAsset).Result;

                Assert.AreEqual((ulong)uploadFileSize, downloadResult.GetDataLength());
            }
        }
#endif

    }
}

File: Tests/Web/html/tests/settingssystem_tests.js
import { test, assert } from '../test_framework.js';
import { generateUniqueString } from '../test_helpers.js';
import { logIn } from './usersystem_tests_helpers.js';
import { createSpace } from './spacesystem_tests_helpers.js';

import { freeBuffer, Systems } from '../connected_spaces_platform.js';
import { createBufferAssetDataSource} from './assetsystem_tests_helpers.js'


test('SettingsSystemTests', 'BlockSpaceTest', async function() {
    const systemsManager = Systems.SystemsManager.get();
    const userSystem = systemsManager.getUserSystem();
    const spaceSystem = systemsManager.getSpaceSystem();
    const settingsSystem = systemsManager.getSettingsSystem();

    const spaceName = generateUniqueString('CSP-TESTS-WASM-SPACE');
    const spaceDescription = 'CSP-TESTS-WASM-SPACEDESC';

    // Log in
    const userId = await logIn(userSystem);

    // Create space
    const space = await createSpace(spaceSystem, spaceName, spaceDescription, Systems.SpaceAttributes.Private);

    // Clear blocked spaces
    {
        const result = await settingsSystem.clearBlockedSpaces(userId);

        assert.succeeded(result);

        result.delete();
    }

    // Block space
    {
        const result = await settingsSystem.addBlockedSpace(userId, space.id);

        assert.succeeded(result);

        result.delete();
    }

    // Get blocked spaces
    {
        const result = await settingsSystem.getBlockedSpaces(userId);

        assert.succeeded(result);

        const spaceIds = result.getValue();
        result.delete();

        assert.areEqual(spaceIds.size(), 1);

        const spaceId = spaceIds.get(0);
        spaceIds.delete();

        assert.areEqual(spaceId, space.id);
    }
});


test('SettingsSystemTests', 'UnBlockSpaceTest', async function() {
    const systemsManager = Systems.SystemsManager.get();
    const userSystem = systemsManager.getUserSystem();
    const spaceSystem = systemsManager.getSpaceSystem();
    const settingsSystem = systemsManager.getSettingsSystem();

    const spaceName = generateUniqueString('CSP-TESTS-WASM-SPACE');
    const spaceDescription = 'CSP-TESTS-WASM-SPACEDESC';

    // Log in
    const userId = await logIn(userSystem);

    // Create space
    const space = await createSpace(spaceSystem, spaceName, spaceDescription, Systems.SpaceAttributes.Private);

    // Clear blocked spaces
    {
        const result = await settingsSystem.clearBlockedSpaces(userId);

        assert.succeeded(result);

        result.delete();
    }

    // Block space
    {
        const result = await settingsSystem.addBlockedSpace(userId, space.id);

        assert.succeeded(result);

        result.delete();
    }

    // Get blocked spaces
    {
        const result = await settingsSystem.getBlockedSpaces(userId);

        assert.succeeded(result);

        const spaceIds = result.getValue();
        result.delete();

        assert.areEqual(spaceIds.size(), 1);

        const spaceId = spaceIds.get(0);
        spaceIds.delete();

        assert.areEqual(spaceId, space.id);
    }

    // Unblock space
    {
        const result = await settingsSystem.removeBlockedSpace(userId, space.id);

        assert.succeeded(result);

        result.delete();
    }

    // Get blocked spaces
    {
        const result = await settingsSystem.getBlockedSpaces(userId);

        assert.succeeded(result);

        const spaceIds = result.getValue();
        result.delete();

        assert.areEqual(spaceIds.size(), 0);

        spaceIds.delete();
    }
});

test('SettingsSystemTests', 'UpdateAvatarPortraitWithBuffer', async function() {
    const systemsManager = Systems.SystemsManager.get();
    const userSystem = systemsManager.getUserSystem();
    const assetSystem = systemsManager.getAssetSystem();
    const settingsSystem = systemsManager.getSettingsSystem();


    // Log in
    var userID = await logIn(userSystem);

    {
        // Create Avatar Portrait
        var avatarPortrait = await createBufferAssetDataSource("/assets/OKO.png","image/png");

        const result = await settingsSystem.updateAvatarPortraitWithBuffer(userID, avatarPortrait);
        freeBuffer(avatarPortrait.buffer);

        assert.succeeded(result);

        let avatarPortraitUri = "";

        // Get Avatar Portrait
        {
            const result = await settingsSystem.getAvatarPortrait(userID);

            assert.succeeded(result);

            avatarPortraitUri = result.getUri();
            result.delete();
        }
        // Create asset
        let testAsset = Systems.Asset.create();
        testAsset.uri = avatarPortraitUri;

        // Download current Avatar Portrait data
        {
            const result = await assetSystem.downloadAssetData(testAsset);

            assert.succeeded(result);

            // Check data lengths
            const dataLength =  result.getDataLength();
            result.delete();
            console.log("New Thumbnail From Space Data Length: " + String(dataLength) + " bytes");

            assert.areEqual(dataLength,avatarPortrait.bufferLength);
        }
        avatarPortrait.delete();
        testAsset.delete();
    }
});